### DISCUSSION TITLE
API to call from PHP to Golang and Vice Versa?

### DISCUSSION THREAD

**Discussion by skurfuerst** [php/frankenphp#66] [Ideas]
Hey,

I am thoroughly impressed by FrankenPHP - especially the relatively little amount of code for such a huge functionality.

I plan to try it out with NeosCMS really soon 😀 

A related question: do you plan to add a general "call_go_func" from within PHP code (and maybe also the other way around?) that would imho be huge, and I imagine it would be quite little work.

Would be curious to hear your thoughts about it :)

All the best,
Sebastian

PS out of curiosity, how many changes were needed on plain PHP? I somehow could not find the diff you added to PHP 8.2 to make all this work.

**Comment by dunglas** [comment]
Good idea! It's actually already possible (users just have to provide their own extensions). This "just" has to be documented.
We could probably document how to integrate https://github.com/arnaud-lb/php-go from @arnaud-lb too.

> PS out of curiosity, how many changes were needed on plain PHP? I somehow could not find the diff you added to PHP 8.2 to make all this work.

The latest version (`main` branch) works with vanilla PHP 8.2 (dev branch) as long as it is compiled with `--enable-zts` and `--disable-zend-signals`.

**Reply by skurfuerst** [reply]
Hey,

Really cool :):) thanks for the hint, did not know about the other package yet.

All the best, and keep up the awesome work,
Sebastian

**Comment by zaunere** [comment]
Hi,

Just got frankenphp online and it's looking pretty incredible, and it's great to have a clean backend.

I guess similar to the OP, can go libs be used in PHP?  I'm going to check out the php-go project, but wondering how it could be done, if possible, in franken itself?  If not, and php-go has to be used, anyone have any implementation/integration ideas?

It'd be incredible if PHP could call the go ecosystem, use libs, goroutines? async,?

And with a modern engine and so lightweight, awesome!

I saw these benchmarks:
https://dev.to/dimdev/performance-benchmark-of-php-runtimes-2lmc

Which led me to this project and I'm on pins and needles! :)

Best,

Hans

**Comment by nickchomey** [comment]
Discussion #1022 also has some information on this.

It really would be great to get some official documentation on it - calling php and go from each other.

**Reply by zaunere** [reply]
I'd be happy to help, but need a starting place to dig up the info - not really anything I've been able to find.

**Comment by leocavalcante** [comment]
Sorry to bother about this again and yes, I know that comparing solutions randomly like this is counter-productive, but... [RR's customization features](https://docs.roadrunner.dev/docs/customization/plugin) are killer ones, at least for me, to choose between one instead of another

**Reply by dunglas** [reply]
You can achieve the same thing as this (and much more) by writing a Caddy module: https://caddyserver.com/docs/extending-caddy

As for RR, if a middleware-like approach isn't enough, you can expose a local HTTP API (or any other protocol you want) to call from your PHP code.

It's also possible to register real PHP functions (it's how FrankenPHP exposes its own functions to PHP: https://github.com/dunglas/frankenphp/blob/main/frankenphp.c#L417), but this requires writing a bit of C and is not documented yet.
